### ISSUE TITLE
Deadlock deleting 2 vertices in parallel

### ISSUE THREAD

**Issue by mbbeaubi** [JanusGraph/janusgraph#1333]
To reproduce, have two vertices with an edge between them. Open 1 transaction. Create two threads. Have the first thread delete the first vertex while at the same time have the second thread delete the second vertex. This is a timing issue so you may need to set some breakpoints to get it to happen. Below is a thread dump of the issue happening in the wild:

```
"propagation--5" #113563 daemon prio=5 os_prio=0 tid=0x00007f93ac8fb000 nid=0x14186 waiting for monitor entry [0x00007f9351768000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.janusgraph.graphdb.vertices.StandardVertex.updateLifeCycle(StandardVertex.java:45)
	- waiting to lock <0x000000073e4a1438> (a org.janusgraph.graphdb.vertices.CacheVertex)
	at org.janusgraph.graphdb.vertices.StandardVertex.removeRelation(StandardVertex.java:51)
	at org.janusgraph.graphdb.transaction.StandardJanusGraphTx.removeRelation(StandardJanusGraphTx.java:611)
	at org.janusgraph.graphdb.relations.CacheEdge.remove(CacheEdge.java:149)
	at org.janusgraph.graphdb.transaction.RelationConstructor$1$1.remove(RelationConstructor.java:67)
	at org.janusgraph.graphdb.vertices.AbstractVertex.remove(AbstractVertex.java:102)
	- locked <0x000000073e48f5e0> (a org.janusgraph.graphdb.vertices.CacheVertex)
	at org.janusgraph.graphdb.vertices.StandardVertex.remove(StandardVertex.java:101)
	- locked <0x000000073e48f5e0> (a org.janusgraph.graphdb.vertices.CacheVertex)
	at com.infa.products.ldm.janusgraph.nextgen.DeleteVertexHandler.handle(DeleteVertexHandler.java:35)
	at com.infa.products.ldm.janusgraph.nextgen.PostIngestHandler$HandlerCallable.call(PostIngestHandler.java:71)
	at com.infa.products.ldm.janusgraph.nextgen.PostIngestHandler$HandlerCallable.call(PostIngestHandler.java:1)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

"propagation--4" #113562 daemon prio=5 os_prio=0 tid=0x00007f93ac799800 nid=0x14185 waiting for monitor entry [0x00007f935196a000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.janusgraph.graphdb.vertices.StandardVertex.updateLifeCycle(StandardVertex.java:45)
	- waiting to lock <0x000000073e48f5e0> (a org.janusgraph.graphdb.vertices.CacheVertex)
	at org.janusgraph.graphdb.vertices.StandardVertex.removeRelation(StandardVertex.java:51)
	at org.janusgraph.graphdb.transaction.StandardJanusGraphTx.removeRelation(StandardJanusGraphTx.java:611)
	at org.janusgraph.graphdb.relations.CacheEdge.remove(CacheEdge.java:149)
	at org.janusgraph.graphdb.transaction.RelationConstructor$1$1.remove(RelationConstructor.java:67)
	at org.janusgraph.graphdb.vertices.AbstractVertex.remove(AbstractVertex.java:102)
	- locked <0x000000073e4a1438> (a org.janusgraph.graphdb.vertices.CacheVertex)
	at org.janusgraph.graphdb.vertices.StandardVertex.remove(StandardVertex.java:101)
	- locked <0x000000073e4a1438> (a org.janusgraph.graphdb.vertices.CacheVertex)
	at com.infa.products.ldm.janusgraph.nextgen.DeleteVertexHandler.handle(DeleteVertexHandler.java:35)
	at com.infa.products.ldm.janusgraph.nextgen.PostIngestHandler$HandlerCallable.call(PostIngestHandler.java:71)
	at com.infa.products.ldm.janusgraph.nextgen.PostIngestHandler$HandlerCallable.call(PostIngestHandler.java:1)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by mbbeaubi** [comment]
The issue happens because of the synchronized around the updateLifeCycle and remove methods in StandardVertex. The locking isn't correct for lifecycle anyways. I'll submit a pull request.

**Comment by porunov** [comment]
Fixed by #1380
